### PR TITLE
Include active trackfiles in backup (rel. to #14706)

### PR DESCRIPTION
**Pull request by moving-bits**
## Description
Includes active track files into database backup/restore.
"active" meaning: those, which are selected for display on the map, disregarding their enabled/disabled state.
(Contents of configured GPX folder are not backed up by this feature.)

Backup/restore is performed when database gets backed up or restored, no separate selection for track files provided. (This is intentionally, as "active" track files are referenced in the database, and this data should be kept in sync.)

### CODE REVIEW

File: main/src/main/java/cgeo/geocaching/utils/BackupUtils.java [hunk 1]
```diff
@@ -300,7 +305,19 @@ public void restoreInternal(final Activity activityContext, final Folder backupD
         };
 
         if (database) {
-            restoreDatabaseInternal(backupDir, consumer);
+            final Folder backupTracksDir = Folder.fromFolder(backupDir, TRACKS_SUBFOLDER);
+            final List<ContentStorage.FileInformation> tracks = ContentStorage.get().list(backupTracksDir);
+            AndroidRxUtils.andThenOnUi(Schedulers.io(), () -> {
+                    final AtomicInteger tfCopied = new AtomicInteger(0);
+                    for (ContentStorage.FileInformation fi : tracks) {
+                        final Uri uri = ContentStorage.get().copy(fi.uri, Folder.fromFile(LocalStorage.getTrackfilesDir()), FileNameCreator.forName(fi.name), false);
```
Comment (eddiemuc) [inline main/src/main/java/cgeo/geocaching/utils/BackupUtils.java:313]: In class FolderUtils you will find methods to copy multiple files/folders at once

**Comment by eddiemuc** [comment]
What about following scenario:
* User performs backup -> tracks are copied to backup dir
* User uninstalls c:geo -> internal dir is cleared, but his public "track" folder is left untouched (track files are still in it)
* User installs c:geo and restores backup -> backup track files are copied into the public "track" folder and hence doubled-up

Wouldn't it be easier just to check the public track folder (and re-copy to internal app dir) if track file is not found in app-internal dir?

File: main/src/main/java/cgeo/geocaching/utils/BackupUtils.java (hunk 1 above)
Comment (moving-bits) [reply main/src/main/java/cgeo/geocaching/utils/BackupUtils.java:313]: Thanks for the pointer, I'll look into it.

**Comment by moving-bits** [comment]
> What about following scenario:
> * User performs backup -> tracks are copied to backup dir
> * User uninstalls c:geo -> internal dir is cleared, but his public "track" folder is left untouched (track files are still in it)
> * User installs c:geo and restores backup -> backup track files are copied into the public "track" folder and hence doubled-up
>    Or is that a false assesment?
> 
> Wouldn't it be easier just to check the public track folder (and re-copy to internal app dir) if track file is not found in app-internal dir?

Track files, which a user select for display on the map, will be copied into an app-internal folder (to make them accessible using `File`). You can delete the track's source, and displaying the track will still work. Also, there's no longer a connection between the source file and the app-internal copy, c:geo would not know whether the source is still available. Thus I think it should be included in the backup. (Typically it's just a handful of files, if at all, and those files having just a few kb in size, so it does not inflate the backup that much.)

File: main/src/main/java/cgeo/geocaching/utils/BackupUtils.java (hunk 1 above)
Comment (moving-bits) [reply main/src/main/java/cgeo/geocaching/utils/BackupUtils.java]: updated PR accordingly in both backup and restore methods

**Comment by eddiemuc** [comment]
> > What about following scenario:
> > 
> > * User performs backup -> tracks are copied to backup dir
> > * User uninstalls c:geo -> internal dir is cleared, but his public "track" folder is left untouched (track files are still in it)
> > * User installs c:geo and restores backup -> backup track files are copied into the public "track" folder and hence doubled-up
> >   Or is that a false assesment?

Did you check that no "double-up" of files happens in the public folder in the above scenario? Or (if it does) is this an accepted side-effect?

**Comment by moving-bits** [comment]
> Did you check that no "double-up" of files happens in the public folder in the above scenario? Or (if it does) is this an accepted side-effect?

backup/restore act with backup folder and app-internal cache folder only, public GPX folder is not involved. Backed-up track files go into `tracks` subfolder of current backup folder, thus no need to check for duplicates here.

There may be duplicates between public GPX folder and app-internal track files cache, but IMHO it's not worth comparing each file in the GPX folder with app-internal track files copy. And if we would leave out those files from backup if they are present in GPX folder (at the time of the backup), backup state would be inconsistent, if you move the backup folder to a different device without copying the GPX folder at the same time.

**Comment by eddiemuc** [comment]
> There may be duplicates between public GPX folder and app-internal track files cache, but IMHO it's not worth comparing each file in the GPX folder with app-internal track files copy. And if we would leave out those files from backup if they are present in GPX folder (at the time of the backup), backup state would be inconsistent, if you move the backup folder to a different device without copying the GPX folder at the same time.

Hm. I don't like it too much, but I won't oppose it.

**Pull request merged**